File: 그리디 부수기/BOJ_31748.java
package BOJ_31748;

import java.io.*;
import java.util.StringTokenizer;

public class BOJ_31748 {

    // https://www.acmicpc.net/problem/31784
    // input
    private static BufferedReader br;

    // variables
    private static int N, K;
    private static String S;
    private static char[] chArr;

    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("C:\\Users\\bigyo\\Desktop\\알고리즘\\JavaAlgorithm\\src\\BOJ_31748\\res.txt"));
        br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        input();

        bw.write(solve());
        bw.close();
    } // End of main()

    private static String solve() {
        StringBuilder sb = new StringBuilder();

        int idx = 0;
        while (K > 0) {
            if (idx == N - 1) {
                // 나머지 처리
                K %= 26;
                if (K > 'Z' + 1 - chArr[idx]) {
                    K -= 'Z' + 1 - chArr[idx];
                    chArr[idx] = 'A';
                }
                chArr[idx] += K;
                K = 0;
            } else {
                if (chArr[idx] == 'A') {
                    idx++;
                    continue;
                } else {
                    int count = 'Z' - chArr[idx] + 1;
                    if (count <= K) {
                        chArr[idx] = 'A';
                        K -= count;
                        idx++;
                    } else {
                        idx++;
                    }
                }
            }
        }

        for (int i = 0; i < N; i++) {
            sb.append(chArr[i]);
        }

        return sb.toString();
    } // End of solve()

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        S = br.readLine();
        chArr = S.toCharArray();
    } // End of input()
} // End of Main class

File: 그리디 부수기/BOJ 31748.kt
package BOJ_31748

import java.io.BufferedReader
import java.io.File
import java.util.StringTokenizer

// input
private var br = System.`in`.bufferedReader()

// variables
private var N = 0
private var K = 0
private lateinit var chArr: CharArray

fun main() {
    val path = "C:\\Users\\bigyo\\Desktop\\알고리즘\\KotlinAlgo\\src\\main\\kotlin\\BOJ_31748\\res.txt"
    br = BufferedReader(File(path).bufferedReader())
    val bw = System.out.bufferedWriter()

    input()

    bw.write(solve())
    bw.close()
} // End of main()

private fun solve(): String {
    val sb = StringBuilder()

    // 앞쪽을 최대한 A로 만들고 나머지를 뒤에서 처리하는 방식

    var idx = 0
    while (K > 0) {
        if (idx == N - 1) {
            K %= 26
            if (K > ('Z' + 1) - chArr[idx]) {
                // A로 만들 수 있으면 회전해서 A에서 시작
                K -= ('Z' + 1) - chArr[idx]
                chArr[idx] = 'A'
            }

            // 나머지 K횟수 활용
            chArr[idx] = (chArr[idx] + K).toChar()
            K = 0 // K 모두 소진 처리
        } else {
            if (chArr[idx] == 'A') {
                idx++
                continue
            }

            val count = ('Z' + 1) - chArr[idx]
            if (count <= K) {
                chArr[idx] = 'A'
                K -= count
                idx++
            } else {
                idx++
            }
        }
    }

    chArr.forEach {
        sb.append(it)
    }

    return sb.toString()
} // End of solve()

private fun input() {
    val st = StringTokenizer(br.readLine())
    N = st.nextToken().toInt()
    K = st.nextToken().toInt()
    chArr = br.readLine().toCharArray()
} // End of input()
